Add Pyroscope Adhoc blog post (#34)
pyroscope-io/docs: blog/pyroscope-adhoc-profiling.mdx

slug: /pyroscope-adhoc-profiling
date: "2022-01-26"

authors:
  - name: Ryan Perry
    url: https://github.com/Rperry2174
    image_url: https://avatars.githubusercontent.com/u/23323466?v=4

tags: [observability, flamegraphs, profiling, go]
---

![image](https://user-images.githubusercontent.com/23323466/151271223-f308890a-8e32-42f6-ab2a-b78178e869da.png)

### When to use Continuous vs Adhoc Profiling

Historically, most profilers have been built for more static or adhoc analysis (i.e. profiling a script) as opposed to Pyroscope's continuous profiling which added the dimension of time. This allows people to do a more continuous analysis (i.e. profiling a server) and jump around to any point in time. 

Over time as we continued to improve on the UI/UX of Pyroscope and [dogfooding our own product](/blog/profile-continuous-profiler) we started to realize that there are a lot of situations where its actually ideal to use profiling in this static manner vs running a profiler continuously. 

- **Profiling servers**: Continuous profiling
- **Profiling scripts**: Adhoc profiling
- **Attaching to a running process**: Adhoc profiling

> Our goal is to make Pyroscope a one-stop-shop for *all* profiling needs... And that means supporting all languages, all flamegraph formats, continuously profiling a servers, and -- the topic of this post -- quickly profiling an adhoc script.

Reviewer: ```suggestion
Our goal is to make Pyroscope a one-stop-shop for *all* profiling needs... And that means supporting all languages, all flamegraph formats, continuously profiling servers, and — the topic of this post — quickly profiling an adhoc script.
```

Reviewer: maybe even:
```suggestion
Our goal is to make Pyroscope a one-stop-shop for *all* profiling needs... And that means supporting all languages, all flamegraph formats, continuously profiling servers, and — the topic of this post — quickly profiling a script.
```

because I don't know if anyone necessarily knows what and adhoc script is

### Introducing Adhoc profiling

> That being said, we are excited to officially release Adhoc profiling mode for Pyroscope! With this mode you get all the convenience and simplicity of profiling a script and also get the visualization/UI functionality that we’ve built out over the life of Pyroscope.

Reviewer: I thought "That being said" means "however," or, "despite what I just said,". So I was expecting you to talk about continuous profiling here.. might just be me

Author: Its more like "to continue speaking about the thing i was speaking before..."

### Exporting interactive Flamegraph HTML from Pyroscope Adhoc mode

You can use the `pyroscope adhoc` command to profile a script and export an HTML file which contains an interactive profile that you can view, play with, and share with your team. 

> For example, in Go to profile a script you simply run:

Reviewer: ```suggestion
For example, to profile a script in Go you simply run:
```

```sh

> pyroscope adhoc go run adhoc-push.go

Reviewer: It's going to be slightly more complex though, we might have to use adhoc pull mode here.

Author: ![Uploading image.png…]()

Reviewer: yeah but right now you have to add a bunch of code to your app to make it work. This snippet to me suggests that it "just works" if I prepend `go run` with `pyroscope adhoc`

Author: added link to the file

```
which will produce a folder containing a `.html` file with an interactive flamegraph (for each of the pprof profiles)

> ![image](https://user-images.githubusercontent.com/23323466/151047215-4cce89a1-4403-48c6-92b0-b6f1291af7c5.png)

Reviewer: ```suggestion

![image](https://user-images.githubusercontent.com/23323466/151047215-4cce89a1-4403-48c6-92b0-b6f1291af7c5.png)
```

This comand will profile the file passed in and then export an html file which looks like the Pyroscope UI, but is **exported as a simple standalone HTML file**:
![image](https://user-images.githubusercontent.com/23323466/151047638-f79db7c0-945e-4862-9c34-1654c89782fa.png)

> Note: This html export can also be achieved through the file exported dropdown available in the flamegraph toolbar:

Reviewer: I don't think it's clear that you're talking about the regular pyroscope server here.

> ![image](https://user-images.githubusercontent.com/23323466/151273431-701f0fe4-4031-467e-8c13-189482771493.png)

Reviewer: just a new line:
```suggestion

![image](https://user-images.githubusercontent.com/23323466/151273431-701f0fe4-4031-467e-8c13-189482771493.png)
```

> This is the simplest way to use Pyroscope adhoc mode and is great for quickly viewing and sharing profiles that are richer than a simple screenshot. 

Reviewer: > that are richer than a simple screenshot

This to me sounds like "this is great for quickly viewing and sharing profiles", but only the **profiles** that are "richer than a simple screenshot". Maybe rephrase it?

### Analyzing Pyroscope-generated Adhoc Flamegraphs

> While standalone mode is great for many use cases, one of the benefits of running a Pyroscope server is that it also has a built in mechanisms for storage that allow you to store these exported flamegraphs and revisit them later.

Reviewer: > While standalone mode is great for many use case, ...

This to me sounds like you're about to say that standalone mode is not great for something, but instead you talk about running `pyroscope server`. Also, I don't think it's clear what this "standalone" mode is. I would drop standalone mode, and maybe rephrase the whole paragraph, something like:

```suggestion
HTML file export allows you to easily share profiles with other people. In addition to that pyroscope permanently stores all of your profiles in the `~/.pyroscope/` directory. This allows you to revisit these profiles later or even compare and / or diff them. In order to do that run:
```

Pyroscope stores data in the `~/.pyroscope/` directory. If you run:
```
pyroscope server
```
And go to port `:4040`, you will see a web interface that you can use to view and interact with the flamegraphs you have created via the `pyroscope adhoc` command.

> ![2022-01-25 11 55 01](https://user-images.githubusercontent.com/23323466/151049856-f2388755-97ca-40e7-9af7-ba6f374f1b89.gif)

Reviewer: just a new line:
```suggestion

![2022-01-25 11 55 01](https://user-images.githubusercontent.com/23323466/151049856-f2388755-97ca-40e7-9af7-ba6f374f1b89.gif)
```

### Viewing Flamgraph diffs between two adhoc profiles
Similar to how we have diff view for continuous profiling. Adhoc also supports calculating the diff between two Flamegraphs. 

> For now this functionality requires server-side code so it is only available for files saved inthe `~/.pyroscope` directory. However, we plan to improve on this over time (or potentially even compile these functions to [WASM](https://github.com/pyroscope-io/pyroscope/pull/571) — but thats another post for a nother time).

Reviewer: ```suggestion
For now this functionality requires server-side code so it is only available for files saved inthe `~/.pyroscope` directory. However, we plan to improve on this over time (or potentially even compile these functions to [WASM](https://github.com/pyroscope-io/pyroscope/pull/571) — but thats another post for another time).
```

> ![image](https://user-images.githubusercontent.com/23323466/151269754-fb1f8c1c-8a0b-48bc-9566-e860e7e08f8c.png)

Reviewer: just a new line:
```suggestion

![image](https://user-images.githubusercontent.com/23323466/151269754-fb1f8c1c-8a0b-48bc-9566-e860e7e08f8c.png)
```

### Coming soon!

> As of right now (Jan 26), this mode is mostly for three main use cases:

Reviewer: ```suggestion
As of right now (Jan 26), adhoc mode is meant for three main use cases:
```

> - Profiling scripts using pyroscope: Running `pyroscope adhoc` command in some form and being able to view the flamegraphs that save to `~/.pyroscope`

Reviewer: I would make it more concise, and just say that you can profile scripts and share results with friends

- Viewing profiles exported from pyroscope: Dragging and dropping files that have been exported to `json` format and then analyzing those flamegraphs

> - Getting a shareable standalone flamegraph that has the same UI and analysis features as the web interface

Reviewer: ```suggestion
- Getting a shareable standalone HTML flamegraph that has the same UI and analysis features as pyroscope web interface
```

> However, over time we plan to expand this so that we can support more langauges formats and use cases:

Reviewer: ```suggestion
However, over time we plan to expand functionality so that we can support more languages, formats and use cases:
```

- Supporting drag-and-drop for pprof files
- Adding ability to comment on stored files
- Adding descriptions / annotations to flamegraphs
- Getting a shareable link to flamegraphs
- UI improvements
- and many more! 

> Let us know if you have any recommendations for how to improve this mode!

Reviewer: ```suggestion
Let us know if you have any recommendations for how to improve adhoc mode!
```
